packages/block-editor/src/components/block-pattern-setup/index.js

```diff
@@ -0,0 +1,187 @@
+				</ul>
+			</div>
+		);
+	} else if ( viewMode === VIEWMODES.grid ) {
+		content = (
+			<Composite
+				{ ...composite }
+				role="listbox"
+				className={ containerClass }
+				aria-label={ __( 'Patterns list' ) }
+			>
+				{ patterns.map( ( pattern ) => (
+					<BlockPattern
+						key={ pattern.name }
+						pattern={ pattern }
+						onSelect={ onBlockPatternSelect }
+						composite={ composite }
+					/>
+				) ) }
+			</Composite>
+		);
+	}
+	return content;
+};
+
+function BlockPattern( { pattern, onSelect, composite } ) {
+	const baseClassName = 'block-editor-block-pattern-setup-list';
+	const { blocks, title, description, viewportWidth = 700 } = pattern;
+	const descriptionId = useInstanceId(
+		BlockPattern,
+		`${ baseClassName }__item-description`
+	);
+	return (
+		<div
+			className={ `${ baseClassName }__list-item` }
+			aria-label={ pattern.title }
+			aria-describedby={ pattern.description ? descriptionId : undefined }
+		>
+			<CompositeItem
+				role="option"
+				as="div"
+				{ ...composite }
+				className={ `${ baseClassName }__item` }
+				onClick={ () => onSelect( blocks ) }
+			>
+				<BlockPreview
+					blocks={ blocks }
+					viewportWidth={ viewportWidth }
+				/>
+				<div className={ `${ baseClassName }__item-title` }>
+					{ title }
+				</div>
+			</CompositeItem>
+			{ !! description && (
+				<VisuallyHidden id={ descriptionId }>
+					{ description }
+				</VisuallyHidden>
+			) }
+		</div>
+	);
+}
+
+function BlockPatternSlide( { className, pattern } ) {
+	const { blocks, title, description } = pattern;
+	const descriptionId = useInstanceId(
+		BlockPatternSlide,
+		'block-editor-block-pattern-setup-list__item-description'
+	);
+	return (
+		<li
+			className={ `pattern-slide ${ className }` }
+			aria-label={ title }
+			aria-describedby={ description ? descriptionId : undefined }
+		>
+			<BlockPreview blocks={ blocks } __experimentalLive />
+			{ !! description && (
+				<VisuallyHidden id={ descriptionId }>
+					{ description }
+				</VisuallyHidden>
+			) }
+		</li>
+	);
+}
+
+const BlockPatternSetup = ( {
+	clientId,
+	blockName,
+	filterPatternsFn,
+	startBlankComponent,
+} ) => {
+	const [ viewMode, setViewMode ] = useState( VIEWMODES.carousel );
+	const [ activeSlide, setActiveSlide ] = useState( 0 );
+	const [ showBlank, setShowBlank ] = useState( false );
+	const { replaceBlock } = useDispatch( blockEditorStore );
+	const patterns = usePatternsSetup( clientId, blockName, filterPatternsFn );
+
+	if ( ! patterns?.length || showBlank ) {
+		return startBlankComponent;
+	}
+
+	const onBlockPatternSelect = ( blocks ) => {
+		const clonedBlocks = blocks.map( ( block ) => cloneBlock( block ) );
+		replaceBlock( clientId, clonedBlocks );
+	};
+	return (
+		<div
+			className={ `block-editor-block-pattern-setup view-mode-${ viewMode }` }
+		>
+			<SetupToolbar
+				viewMode={ viewMode }
+				setViewMode={ setViewMode }
+				activeSlide={ activeSlide }
+				totalSlides={ +patterns?.length }
```

Comment:
Clever, but I'm not sure what this is adding to the code. For instance, if `patterns` is undefined, `+patterns?.length` evals to `NaN`, which is not ideal. Are there any circumstances in which `patterns` would **not** be an array? If so, let's address that. But normally we should be able to stick to:

```suggestion
				totalSlides={ patterns.length }
```

Reply:
> +patterns?.length evals to NaN

Missed that... Thanks!!